xds/internal/balancer/gracefulswitch/gracefulswitchbalancer.go

```diff
@@ -0,0 +1,374 @@
+		// this instant (i.e. the LB policy is not ready) you can send it later
+		// on an event like current LB exits READY.
+		gsb.pendingState = state
+		// "If the channel is currently in a state other than READY, the new
+		// policy will be swapped into place immediately." - Java
+		// implementation.
+		if state.ConnectivityState != connectivity.Connecting || !gsb.currentLbIsReady {
+			gsb.swap()
+		}
+	} else {
+		gsb.currentLbIsReady = state.ConnectivityState == connectivity.Ready
+		// In the specific case that the current lb exits ready, and there is a
+		// pending lb, you can forward the pending LB's cached State up to
+		// ClientConn and swap the pending into the current. "Otherwise, the
+		// channel will keep using the old policy until...the old policy exits
+		// READY." - Java
+		if !gsb.currentLbIsReady && gsb.balancerPending != nil {
+			gsb.swap()
+		} else {
+			// Even if there is a pending balancer waiting to be gracefully
+			// switched to, still forward current balancer updates to the Client
+			// Conn. Ignoring state + picker from the current would cause
+			// undefined behavior/cause the system to behave incorrectly from
+			// the current LB policies perspective. Also, the current LB is
+			// still being used by grpc to choose SubConns per RPC, and thus
+			// should use the most updated form of the current balancer.
+			gsb.cc.UpdateState(state)
+		}
+	}
+}
+
+// swap swaps out the current lb with the pending LB and updates the ClientConn.
+// The caller must hold gsb.mu.
+func (gsb *gracefulSwitchBalancer) swap() {
+	gsb.cc.UpdateState(gsb.pendingState)
+	finished := make(chan struct{})
+	go func() {
+		gsb.swapMu.Lock()
+		defer gsb.swapMu.Unlock()
+		gsb.balRecentlyClosed = gsb.balancerCurrent
+		gsb.balancerCurrent.Close()
+		for sc, bal := range gsb.scToSubBalancer {
+			if bal == gsb.balancerCurrent {
+				gsb.cc.RemoveSubConn(sc) // Or does this happen implicitly?
+				delete(gsb.scToSubBalancer, sc)
+			}
+		}
+
+		gsb.balancerCurrent = gsb.balancerPending
+		gsb.balancerPending = nil
+		finished <- struct{}{}
+	}()
+	<-finished
+}
+
+// Helper function that checks if the balancer passed in is current or pending.
+// The caller must hold gsb.mu.
+func (gsb *gracefulSwitchBalancer) balancerCurrentOrPending(bal balancer.Balancer) bool {
+	return bal == gsb.balancerCurrent || bal == gsb.balancerPending
+}
+
+func (gsb *gracefulSwitchBalancer) newSubConn(bal balancer.Balancer, addrs []resolver.Address, opts balancer.NewSubConnOptions) (balancer.SubConn, error) {
+	if gsb.closed.HasFired() {
+		// logger?
+		return nil, errBalancerClosed
+	}
+	gsb.mu.Lock()
+	defer gsb.mu.Unlock()
+	if !gsb.balancerCurrentOrPending(bal) {
+		return nil, fmt.Errorf("%T at address %p that called NewSubConn is deleted", bal, bal)
+	}
+	sc, err := gsb.cc.NewSubConn(addrs, opts)
+	if err != nil {
+		return nil, err
+	}
+	gsb.scToSubBalancer[sc] = bal
+	return sc, nil
+}
+
+// Eat newSubConn calls from current if there is a pending?
```

Comment:
This shouldn't be necessary, as long as you make sure the SubConns are deleted correctly in swap().

Reply:
Sorry, I should have deleted this comment and the comment below. These comments were remants of a thought Doug had earlier, which was if there's a current balancer and a pending balancer, would these be even worth it for the ClientConn to do if there a pending balancer about to be switched to (especially if we add a fixed timer which then switches it automatically after a set time period). However, we all agreed that it would be fine (there is even a comment on this PR where you said should be fine) to not try and do this optimization.

Reply:
Deleted this comment and the one below.